### ISSUE TITLE
how to make embedded links reference "extracted ebook chapter" versus "web page chapter"

### ISSUE THREAD

**Issue by seasaltseasalt** [dteviot/WebToEpub#375]
Hi:
My title has alot of internal links
- references
- end notes
- subject index
- author index

How do have the extract web page to EPUB replace the full URL reference for web page (book chapter) to reference of book only ?  
For example, the embedded link shows are long URL, made of components
https://[domain].[my access proxy].view.[publishers bookID]/[online book chapterID-bibItem-34

book only is
online book chapterID-bibItem-34

Because the extract has kept the full URL in all embedded links, all links take you externally (web page) versus internally (EPUB)

When I extract the web page to EPUB, 2 things are happening
1 When I click on embedded link item in EPUB, I am navigated externally to login (university proxy) browser so can access content. Rather than go to the section same chapter or another chapter in book
2 Online when I click embedded link item for chapter reference item, I am navigated to specific/exact reference item e.g. bib item 34.  This behaviour is lost after extract web page to EPUB. When I extract to EPUB, this is lost as there is no element ID for individual reference items. Some chapters the reference list is over 100 items.  Perhaps somehow I have to tell the tool, to go down a level, I am not sure.  If I use the web page save as PDF function, the link to correct reference item (e.g. bib item 34) works in the PDF. 

Any tips on how to proceed?

Thankyou for your time.

**Comment by dteviot** [comment]
@seasaltseasalt 
 I'm not sure I understand what you're asking for.
I think you're saying something like:
1. Normally, WebToEpub is given a web page with a set of hyperlinks.
2. WebToEpub then fetches each page the links point to and turns them into an epub, with each web page being a chapter
3. That doesn't work for these references, because the rreference takes you to a landing page on the university proxy, and you then need to follow a link from there to the actual web page you want.

If this is correct, then you will need to write a custom parser to handle this.  Instructions for a custom parser are at https://dteviot.github.io/Projects/webToEpub_CustomizingParserTemplate.html.  What you'll need to do is replace the fetchChapter(() function.  When called, it needs to fetch the first URL, and then follow the URL retrieved from the first URL.   Somehting like this:
https://github.com/dteviot/WebToEpub/blob/ae26acf087c86c90e9dfed8a90d1bbe2b43ae51c/plugin/js/parsers/FlyingLinesParser.js#L37-L44

**Comment by seasaltseasalt** [comment]
Thankyou dteviot for your reply.  I am a beginner so I thankyou for your patience.

Is it typical that for each new website, hosting ebooks, that I will need to create a **custom parser**?
If so, I need to get very good at this task. Thankyou for reference.

Each website I use ebooks from, has a distinct layout.  But all the websites have the common elements I am struggling with EMBEDDED LINKS, both internal and external links.  

- **Internal links:** internal to the book (many chapters)
- **External links**: external to the book e.g. external reference source of some sort. Often appear as part of each Chapter Reference bibliographical item for further reading

**1 Internal links** 
**1.1 Chapter Reference Notes** - these are linked in html (if I save as HTML I can see the link). Problem is I am failing to get the link in my extract. I get the text in my extract but no link.
**1.2 Subject Index** - these are linked in the html but the link is to external reference versus an internal reference. How you interpreted the problem above is nearly what I was trying to ask.  
1. Normally, WebToEpub is given a web page with a set of hyperlinks
2. WebToEpub fetches each page and the links (the link will show an external reference as that is what is there). However the external reference, as actually another chapter of same book, that is also available online. As I am getting the chapters offline, I need to make this external reference to an internal reference (another chapter extracted by WebToEpub).  
3 Currently as it is an external reference, when the link is clicked, it will ask me to authenticate. none of that is needed as I have extracted the other chapter, so it needs to go that chapter or part of page of same chapter.

**Comment by seasaltseasalt** [comment]
Thankyou dteviot for your reply.  I am a beginner so I thankyou for your patience.

Is it typical that for each new website, hosting ebooks, that I will need to create a **custom parser**?
If so, I need to get very good at this task. Thankyou for reference.

Each website I use ebooks from, has a distinct layout.  But all the websites have the common elements I am struggling with EMBEDDED LINKS, both internal and external links.  

- **Internal links:** internal to the book (many chapters)
- **External links**: external to the book e.g. external reference source of some sort. Often appear as part of each Chapter Reference bibliographical item for further reading

**1 Internal links** 
**1.1 Chapter Reference Notes** - these are linked in html (if I save as HTML I can see the link). Problem is I am failing to get the link in my extract. I get the text in my extract but no link.
**1.2 Subject Index** - these are linked in the html but the link is to external reference versus an internal reference. How you interpreted the problem above is nearly what I was trying to ask.  
1. Normally, WebToEpub is given a web page with a set of hyperlinks
2. WebToEpub fetches each page and the links (the link will show an external reference as that is what is there). However the external reference, as actually another chapter of same book, that is also available online. As I am getting the chapters offline, I need to make this external reference to an internal reference (another chapter extracted by WebToEpub).  
3 Currently as it is an external reference, when the link is clicked, it will ask me to authenticate. none of that is needed as I have extracted the other chapter, so it needs to go that chapter or part of page of same chapter.

**Comment by dteviot** [comment]
@seasaltseasalt
>Is it typical that for each new website, hosting ebooks, that I will need to create a custom parser?

It depends on how similar they are.

I really need to see some examples so I can see what's going on. Any chance you can send me URLs to pages?
Also, it's getting late for me, and I have other things to do, so I probably won't respond until tomorrow.

**Comment by dteviot** [comment]
@seasaltseasalt
OK, I was having trouble understanding what you meant by internal and external references.
But I think I get it now.

Scenario is as follows.
1. We start with a Table of Contents page, which has a set of hyperlinks.  To keep things simple, lets assume there's two links, call them Chapter 1 and Chapter 2
2. WebToEpub (W2E) then fetches the two pages and makes them into an Epub with two chapters.
3. However, the pages also have Hyperlinks.  Chapter 1 might have 3 types of hyperlinks.
   A. A link to another part of Chapter 1 (e.g. Link to reference)
   B. A link to somewhere in Chapter 2
   C. A link to somewhere that is not any of the chapters in the ToC.

The problem you have is:
Links in case A and B are not being converted into what you call "internal references", that is, links that point to elsewhere in the created Epub.
Instead, the links are left in their original format.  So when you click on the links in the epub, it tries to go to the original source, rather than the copy in the Epub.

I'm somewhat surprised this is happening. W2E is supposed to fix up hyperlinks in the first two cases.
That said, there are multiple ways to write the same URL, so it's possible that W2E is failing to match the URLs and so replace the links.
Trying to determine why it's not working without access to the source might be difficult.

That said, W2E embeds metadata into the EPUB about where it obtained each chapter from, so it might be possible to fix up the links after an EPUB is created.
Oddly enough, I'm currently working on a tool https://github.com/dteviot/MergeWebToEpub to do something rather similar. 
If you can send me an Epub made with W2E, I'll take a look and see what can be done.

Additional, you also said something like.
>  Online when I click embedded link item for chapter reference item, I am navigated to specific/exact reference item e.g. bib item 34. This behaviour is lost after extract web page to EPUB. When I extract to EPUB, this is lost as there is no element ID for individual reference items

i.e. In case A, W2E is removing ANCHORS (the ID attribute in an element).  That is, each reference at the end of the chapter should have an ID attribute that matches the URL after the '#', but this attribute is being removed.  I'm a bit puzzled by that.  I recall the W2E DID do that when it could not match an Anchor with a link (because it upset EPUB validation tools.)  However, my notes suggest I removed that code years ago.  Also, looking at the two HTML sample pages you sent me, I can't find any ID attributes

Anyway, send me an EPUB, and the raw HTML of a matching chapter with A type links and I'll see what I can do.
As regards type C links, I'm assuming you're not asking me to solve them.

**Comment by dteviot** [comment]
@seasaltseasalt
Thanks for you email, I think I've got what I need.
I'm a bit pressed for time tonight (and probably will be for the next few as well)

> Type A is not going to exact location e.g. Click Chapter Reference (bib item 34) in Chapter

That's probably due to how the ebook reader navigates to the Anchor.

>  Type C hyperlinks are still showing mcgill proxy rather than link without the proxy bit.

Well, yes, they will. 

> Is there a simple way to do remove " https://login.proxy3.library.mcgill.ca/login?qurl=" from all these external links  I

Simple if you do a custom parser.  However, unless you're a developer, creating a custom parser proabbly isn't simple.
But my plan is for MergeWebToEpub to do this..

> I had thought I used element ContentRoot not body but that did not work today.

it's a case sensitive ID, so you need to use "#contentRoot", without the quotes.
Likewise, CSS for the title is "#pagetitle"

That said, once you've given the default parser sett tings for a site, it's not easy to change them.
An unfortunate oversight on my part.  I think if you uninstall the extension and then install again, that should reset things and let you set them again.

Finally, looking at the epub you sent me, I'm reasonably sure I can modify MergeWebToEpub to take the epubs generated by WebToEpub and clean them up.
i.e. 
1. Keep the title and contentRoot for each page and throw the rest away
2. Fix the Type C hypterlinks as you request.

However, I probably won't have time to work on this until the weekend.

**Comment by dteviot** [comment]
@seasaltseasalt
I have attempted to add a custom parser to WebToEpub for the www-oxfordclinicalpsych-com.proxy3.library.mcgill.ca site 
Note that as I can't test it, I'm not sure how well it will work (or even if), so you'll need to try it out.
Anyway, you can download it from https://drive.google.com/file/d/1fNUkkvkk6gp_A4SXSkQ4lvCN1Yxv4-ic/view?usp=sharing
Here's how to install it on chrome https://www.cnet.com/how-to/how-to-install-chrome-extensions-manually/
Skip the .crx to .zip step, I've already done that.  (You'll probably also have to uninstall the existing WebToEpub from chrome.)

The actual parser for the site is the file "OxfordClinicalPsychParser.js" in the "parsers" folder

At the moment, it's pretty stupid.
It just grabs the content using "#list-chapters", the title using "#list-chapters" and tries to fix up the images.

If it works, you might like to try modifying it to remove more "garbage".
Basically, line 37, replace "button" with the CSS selector for the elements to remove.

Also, if it works, let me know and I'll see if I can expand it to clean up the type C hyperlinks.

If it doesn't copy/paste the error text and I'll try and figure out what's not working.

**Comment by dteviot** [comment]
@seasaltseasalt
In case you can't get the extension to work on Chrome, here's a link to a version that should work with Firefox
https://drive.google.com/file/d/1lMOQ_8-IBWEWUk4kUp7THG9oIs5eCpiG/view?usp=sharing
Note, the extensions for Chrome and Firefox are not the same.

To install on Firefox
1. Unpack the zip
2. Open Firefox and type "about:addons" into the URL bar (don't include the quotes)
3. Click "Load Temporary Add-on".
4. Select "Extensions"
5. Click the icon that looks like a gear that is on the right of "Manage Your Extensions"
6. Click "Debug Add-ons"
7. Click "Load Temporary Add-on"
8. Select "manifest.json" from the unpacked zip.

**Comment by seasaltseasalt** [comment]
@dteviot
successfully loaded on chrome
entered starting URL
click web to epub icon

error returned:


OK

> TypeError: Cannot read property 'querySelectorAll' of null
>     at OxfordClinicalPsychParser.fixupImages (chrome-extension://kbmhfehpflnmnncklpeffaeifaehgiah/js/parsers/OxfordClinicalPsychParser.js:22:30)
>     at OxfordClinicalPsychParser.findContent (chrome-extension://kbmhfehpflnmnncklpeffaeifaehgiah/js/parsers/OxfordClinicalPsychParser.js:16:14)
>     at OxfordClinicalPsychParser.findCoverImageUrl (chrome-extension://kbmhfehpflnmnncklpeffaeifaehgiah/js/Parser.js:129:32)
>     at OxfordClinicalPsychParser.populateUI (chrome-extension://kbmhfehpflnmnncklpeffaeifaehgiah/js/Parser.js:120:29)
>     at processInitialHtml (chrome-extension://kbmhfehpflnmnncklpeffaeifaehgiah/js/main.js:45:24)
>     at populateControlsWithDom (chrome-extension://kbmhfehpflnmnncklpeffaeifaehgiah/js/main.js:211:9)
>     at onMessageListener (chrome-extension://kbmhfehpflnmnncklpeffaeifaehgiah/js/main.js:16:13)

**Comment by seasaltseasalt** [comment]
@dteviot
successfully loaded on chrome
entered starting URL
click web to epub icon

error returned:


OK

> TypeError: Cannot read property 'querySelectorAll' of null
>     at OxfordClinicalPsychParser.fixupImages (chrome-extension://kbmhfehpflnmnncklpeffaeifaehgiah/js/parsers/OxfordClinicalPsychParser.js:22:30)
>     at OxfordClinicalPsychParser.findContent (chrome-extension://kbmhfehpflnmnncklpeffaeifaehgiah/js/parsers/OxfordClinicalPsychParser.js:16:14)
>     at OxfordClinicalPsychParser.findCoverImageUrl (chrome-extension://kbmhfehpflnmnncklpeffaeifaehgiah/js/Parser.js:129:32)
>     at OxfordClinicalPsychParser.populateUI (chrome-extension://kbmhfehpflnmnncklpeffaeifaehgiah/js/Parser.js:120:29)
>     at processInitialHtml (chrome-extension://kbmhfehpflnmnncklpeffaeifaehgiah/js/main.js:45:24)
>     at populateControlsWithDom (chrome-extension://kbmhfehpflnmnncklpeffaeifaehgiah/js/main.js:211:9)
>     at onMessageListener (chrome-extension://kbmhfehpflnmnncklpeffaeifaehgiah/js/main.js:16:13)

**Comment by seasaltseasalt** [comment]
@dteviot
successfully loaded on chrome
entered starting URL
click web to epub icon

error returned:

TypeError: Cannot read property 'querySelectorAll' of null
    at OxfordClinicalPsychParser.fixupImages (chrome-extension://kbmhfehpflnmnncklpeffaeifaehgiah/js/parsers/OxfordClinicalPsychParser.js:22:30)
    at OxfordClinicalPsychParser.findContent (chrome-extension://kbmhfehpflnmnncklpeffaeifaehgiah/js/parsers/OxfordClinicalPsychParser.js:16:14)
    at OxfordClinicalPsychParser.findCoverImageUrl (chrome-extension://kbmhfehpflnmnncklpeffaeifaehgiah/js/Parser.js:129:32)
    at OxfordClinicalPsychParser.populateUI (chrome-extension://kbmhfehpflnmnncklpeffaeifaehgiah/js/Parser.js:120:29)
    at processInitialHtml (chrome-extension://kbmhfehpflnmnncklpeffaeifaehgiah/js/main.js:45:24)
    at populateControlsWithDom (chrome-extension://kbmhfehpflnmnncklpeffaeifaehgiah/js/main.js:211:9)
    at onMessageListener (chrome-extension://kbmhfehpflnmnncklpeffaeifaehgiah/js/main.js:16:13)

**Comment by seasaltseasalt** [comment]
Is there a way I can get WebToEpub (WTE) tool to use a simpler naming convention, of the chapters (Text element) in the epub?

For example **Text** 

0003_1__Theoretical_background     _would be_     11_chapter1.xhtml      
0017_Appendix_I__STQ_items_and_normative_data   _would be_    30_appendix-1
0020_Author_index                                                       _would be_    33_indexList-1

(sort similar structure as used on the web page id's (e.g. m-ISBN-appendix-2-p-209)

Being very new to CSS and HTML, I find it simpler to follow when editting the ePUB.

**Comment by dteviot** [comment]
OK, screw-up on my part.
In my defense, I wasn't able to test it before sending it to you.
Anyway, I think I've fixed that problem.
I've created a shared Google Drive folder and put fixed builds of both
there.
https://drive.google.com/drive/folders/1B_X2WcsaI_eg9yA-5bHJb8VeTZGKExl8?usp=sharing

> Is there a way I can get WebToEpub (WTE) tool to use a simpler naming
convention, of the chapters (Text element) in the epub?

Sorry.  No there isn't.

>
>
